Allow tests to filter on availability of full C library

keith-packard opened the pull request:
Some tests are filtering on `not CONFIG_MINIMAL_LIBC` and others on `TOOLCHAIN_HAS_NEWLIB || CONFIG_PICOLIBC_SUPPORTED`, both of which have issues in correctly identifying the desired targets. Instead, create a new hidden Kconfig value, `SUPPORT_FULL_LIBC`, which currently depends on `NATIVE_APPLICATION || TOOLCHAIN_HAS_NEWLIB`.

When picolibc is added as a default in `LIBC_IMPLEMENTATION` , `SUPPORT_FULL_LIBC` can be updated which will automagically allow those tests to run using picolibc if available.

keith-packard committed:
samples/logging/syst: Use FULL_LIBC_SUPPORTED helper Kconfig symbol

This symbol detects whether there is any available full libc for a
particular target, allowing tests to filter on this condition.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
lib/libc: Add FULL_LIBC_SUPPORTED helper Kconfig symbol

This symbol is selected when the target has any full libc available. This
allows tests to filter on this condition. It doesn't depend on whether the
application actually selects that library, only whether requesting a full C
library would work.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
lib/libc: newlib use depends upon newlib being supported

Before allowing newlib to be selected as the C library, ensure that it is
available for the target.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
samples/chre: Filter out targets not providing a full libc

When the only C library available is the minimal one, this test cannot
work. Accept only platforms with full C library support to prevent build
failures.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
tests/subsys/logging: Require a full C library

These tests all require a full C library to run. Replace the condition
checking for either newlib or picolibc with a check for
CONFIG_FULL_LIBC_SUPPORTED instead, then add CONFIG_REQUIRES_FULL_LIBC=y to
ensure a full libc is used.

Signed-off-by: Keith Packard <keithp@keithp.com>

yperess commented on lib/libc/Kconfig:
```diff
@@ -11,6 +11,14 @@ config REQUIRES_FULL_LIBC
 
+config SUPPORT_FULL_LIBC
+	bool
+	depends on NATIVE_APPLICATION || TOOLCHAIN_HAS_NEWLIB
```
I thought you needed `default y` for this to work. I think this change is turning off all those tests.

keith-packard replied:
oops! I think this can be 'def_bool y' instead of just 'bool'. Will try.

keith-packard replied:
(and, TOOLCHAIN_HAS_NEWLIB isn't a Kconfig variable anyways, so that doesn't work).

keith-packard commented:
Converted to draft; I don't know how to detect NEWLIB support in Kconfig.

keith-packard commented:
Ok, I've added Kconfig symbol `NEWLIB_SUPPORTED` that parallels `PICOLIBC_SUPPORTED`. Of course, this requires a patch to sdk-ng as that includes its own Kconfig bits. https://github.com/zephyrproject-rtos/sdk-ng/pull/626

galak commented on lib/libc/Kconfig:
```diff
@@ -26,6 +34,10 @@ config PICOLIBC_SUPPORTED
 	  Selected when the target has support for picolibc.
 
+config NEWLIB_SUPPORTED
+	def_bool y
```
Doesn't feel like this should be a `def_bool`.

Also, can we drop the `depends on` and just introduce something similar to cmake/toolchain/gnuarmemb/Kconfig

galak replied:
> > Doesn't feel like this should be a `def_bool`.
> 
> That seems to be the pattern when you want to create a new symbol built by testing the state of others, but if you've an alternate preferred syntax, I'm happy to adjust.

Yeah, I think this should just be:
```
config NEWLIB_SUPPORTED
   bool
```

and leave it to other places to select it.

> 
> > Also, can we drop the `depends on` and just introduce something similar to cmake/toolchain/gnuarmemb/Kconfig
> 
> I tried that (placing a Kconfig in cmake/toolchain/zephyr) and it didn't work, perhaps because Zephyr SDK provides it's own Kconfig file? I didn't look into it further, once it failed.

We need to see if we can introduce a Kconfig file on the zephyr side that can co-exist w/the Kconfig from the SDK itself.

keith-packard replied:
Happy to adapt if there's an alternate solution available. Do you want to sort that out before the release, or get this bug fixed and then create a better solution later?

stephanosio replied:
> Yeah, I think this should just be:
> 
> ```
> config NEWLIB_SUPPORTED
>    bool
> ```
> 
> and leave it to other places to select it.

The current approach of making `NEWLIB_LIBC_SUPPORTED default y depends on "$(TOOLCHAIN_HAS_NEWLIB)" = "y"` is necessary for smoothing out the transition from the existing `TOOLCHAIN_HAS_NEWLIB` CMake variable/Kconfig to the new `NEWLIB_LIBC_SUPPORTED` Kconfig; otherwise, it will be a breaking change requiring Zephyr SDK and any out-of-tree toolchains to be aware of and properly configure the new `NEWLIB_LIBC_SUPPORTED` Kconfig.

galak reviewed (dismissed):
Would like to see a bit of change to the Kconfig structure.

stephanosio commented on cmake/toolchain/gnuarmemb/Kconfig:
```diff
@@ -6,3 +6,8 @@
 	def_bool y
 	select HAS_NEWLIB_LIBC_NANO
+
+config NEWLIB_SUPPORTED
```
We already have `TOOLCHAIN_HAS_NEWLIB` CMake variable that gets passed down to Kconfig as `CONFIG_TOOLCHAIN_HAS_NEWLIB`.

It should not be necessary to introduce `NEWLIB_SUPPORTED`.

keith-packard replied:
Ah, I had no idea that was available as a config variable. Thanks!

stephanosio commented on lib/libc/Kconfig:
```diff
@@ -11,6 +11,14 @@ config REQUIRES_FULL_LIBC
 	  Helper symbol to indicate some feature requires a C library implementation
 	  with more functionality than what MINIMAL_LIBC provides
 
+config SUPPORT_FULL_LIBC
```
I think we have some inconsistencies in the naming of the helper symbols indicating the support status of LIBCs."

After the changes in this PR, we have:

* SUPPORT_MINIMAL_LIBC
* SUPPORT_FULL_LIBC
* NEWLIB_SUPPORTED
* PICOLIBC_SUPPORTED

I would like to settle on just one naming convention for these symbols, and my personal preference is to use the `*_SUPPORTED`, such that we end up with:

* MINIMAL_LIBC_SUPPORTED
* NEWLIB_LIBC_SUPPORTED
* PICOLIBC_SUPPORTED
* ARCMWDT_LIBC_SUPPORTED
* FULL_LIBC_SUPPORTED

So, every LIBC type will have an associated `_SUPPORTED` symbol that indicates whether the particular LIBC is supported.

If we do this, I think it makes sense to keep the `NEWLIB_LIBC_SUPPORTED` symbol as `default y, depends on TOOLCHAIN_HAS_NEWLIB` for the sake of generalisation -- by doing this, we do not hard-code the support status of `NEWLIB_LIBC` (or any other LIBC for that matter) as simple `TOOLCHAIN_HAS`; it is foreseeable that it may be necessary to add more dependencies (e.g. support status per arch) for some LIBCs in the future.

We can also replicate this pattern to the C++ library side such that we have `GLIBCXX_LIBCPP_SUPPORTED` and so on (just thinking about the bigger picture, not asking this to be done in this PR; I will do that later).

keith-packard replied:
That does look like a nice clean up.

keith-packard replied:
What about the REQUIRES symbols like `REQUIRES_FULL_LIBC`. By symmetry, should these be `FULL_LIBC_REQUIRED`?

keith-packard committed:
lib/libc: Move helper symbols above C library menu

Restructure the file so that the only elements within 'menu "C Library"'
are the library choices.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
lib/libc: Allow REQUIRES_FULL_LIBC to be selected by applications

Add a prompt to the Kconfig symbol so that applications can select this in
their configuration to guide C library selection away from the minimal C
library.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
lib/libc: Replace SUPPORT_MINIMAL_LIBC with MINIMAL_LIBC_SUPPORTED

Clean up libc-related symbols to use a common pattern.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
lib/libc: Add NEWLIB_LIBC_SUPPORTED

This reflects whether newlib is available in the environment. This
symbol should be used in place of TOOLCHAIN_HAS_NEWLIB.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
cmake: Compute TOOLCHAIN_HAS_NEWLIB at cmake time

Switch from using a generator expression to computing the value with a
conditional so that it is available during the execution of kconfig instead
of only being available while generating the build script.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
cmake/toolchain: Add Kconfig setting for toolchains with newlib

Label espressif, gnuarmemd and xtools toolchains with newlib support using
the Kconfig variable rather than relying on the TOOLCHAIN_HAS_NEWLIB
value.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard commented:
Ok, that took longer than I expected. I've updated this PR to include a fairly long series of commits that slowly migrates the tree to using a consistent set of symbols for all of the libc settings:

 * `FULL_LIBC` -- selected by applications requiring a complete C library, but not caring which one. Equivalent to the old symbol `REQUIRES_FULL_LIBC`, except now applications can explicitly set this value, rather than having to specify a C library explicitly.
 * `FULL_LIBC_SUPPORTED` -- selected when there is *some* complete C library available for the application. Useful for tests and samples that want to use twister filters to limit runs to targets for which a complete C library is available.
 * `MINIMAL_LIBC_SUPPORTED` -- selected when the target has support for the minimal C library. This is almost always set, except when using the armclang toolchain. Equivalent the old symbol `SUPPORT_MINIMAL_LIBC`.
 * `NEWLIB_LIBC_SUPPORTED` -- selected when the target has support for newlib. This is computed from `$(TOOLCHAIN_HAS_NEWLIB)" = "y"` at present, but there's a patch in the series which also sets it explicitly for some toolchains.
 * `PICOLIBC_SUPPORTED` -- unchanged in this series

In working on this, I discovered a few things:

 1. `TOOLCHAIN_HAS_NEWLIB` was trying to be set by a generator expression in `kconfig.cmake`, which doesn't work for `execute_process`. I hacked around this by doing the computation in regular cmake code. I think that's equivalent here?
 2. `FULL_LIBC` *cannot* depend upon `FULL_LIBC_SUPPORTED` when we're using `FULL_LIBC_SUPPORTED` in a twister filter. That's because if the dependency is in place, when twister runs cmake to construct the config file from which it extracts the filter variable values, kconfig fails if the application sets `FULL_LIBC` when none is available.

I think I've got things set up so that existing applications using the old Kconfig variables will still work correctly, so strategic bits of this series could be merged before the release if fixing the errors related to clang with some of the tests and samples is interesting. If not, there's nothing else pressing here; when using gcc, things are pretty much the same.

galak commented:
I'd say rename `NEWLIB_LIBC_SUPPORTED` to just `NEWLIB_SUPPORTED` to be consistent with naming of `PICOLIBC_SUPPORTED`.

stephanosio commented:
> I'd say rename `NEWLIB_LIBC_SUPPORTED` to just `NEWLIB_SUPPORTED` to be consistent with naming of `PICOLIBC_SUPPORTED`.

`NEWLIB_LIBC_SUPPORTED` is named as such because the Kconfig symbol for Newlib is `NEWLIB_LIBC`, not `NEWLIB` (in fact, all the other LIBC Kconfigs are suffixed with `_LIBC`; Picolibc is an exception here.

keith-packard committed:
lib/libc: Mark picolibc as providing full C library support

Select FULL_LIBC_SUPPORTED when picolibc is available.

Add picolibc as a secondary default C library when REQUIRES_FULL_LIBC is
selected. This is necessary as tests gated on FULL_LIBC_SUPPORTED need to
be sure that a full C library will be selected -- if only picolibc is
available, those tests will need to select that.

This should permit use of a picolibc-only crosstool-ng toolchain in
testing.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard commented on lib/libc/Kconfig:
```diff
@@ -24,18 +25,24 @@ config PICOLIBC_SUPPORTED
+	bool
+	select FULL_LIBC
+	help
+	  Legacy symbol that selects FULL_LIBC
```
This isn't relevant as `FULL_LIBC` has been removed as suggested below.

keith-packard commented on lib/libc/Kconfig:
```diff
@@ -31,6 +33,12 @@ config REQUIRES_FULL_LIBC
+	bool
+	select MINIMAL_LIBC_SUPPORTED
+	help
+	  Legacy symbol that selects MINIMAL_LIBC_SUPPORTED
```
Yup, thanks for the reminder about `select DEPRECATED`

keith-packard commented on lib/libc/Kconfig:
```diff
@@ -3,11 +3,12 @@
 
-config REQUIRES_FULL_LIBC
+config FULL_LIBC
+	prompt "Require complete C library"
```
Yeah, this does seem better.

keith-packard commented:
I added one more commit to this PR which will select picolibc as the default C library when a full C library is required and newlib isn't available. Now that crosstool-ng has configurations which provide only picolibc, this will be useful there.

keith-packard commented on lib/libc/Kconfig:
```diff
@@ -10,6 +10,13 @@ config REQUIRES_FULL_LIBC
+	depends on NATIVE_APPLICATION || NEWLIB_LIBC_SUPPORTED
+	help
+	  Selected when the target has support for some full C library that
+	  will be selected when REQUIRES_FULL_LIBC is set
```
That's a much better plan. Thanks.

stephanosio reviewed (approved):
LGTM

stephanosio commented:
@yperess @galak PTAL

stephanosio commented:
@galak ping
